packages/grafana-ui/src/utils/statsCalculator.ts

```diff
@@ -0,0 +1,326 @@
+// Libraries
+import isNumber from 'lodash/isNumber';
+
+import { TableData, NullValueMode } from '../types/index';
+
+export enum StatID {
+  sum = 'sum',
+  max = 'max',
+  min = 'min',
+  logmin = 'logmin',
+  mean = 'mean',
+  last = 'last',
+  first = 'first',
+  count = 'count',
+  range = 'range',
+  diff = 'diff',
+  delta = 'delta',
+  step = 'step',
+
+  allIsZero = 'allIsZero',
+  allIsNull = 'allIsNull',
+}
+
+export interface ColumnStats {
+  [key: string]: any;
+}
+
+// Internal function
+type StatCalculator = (data: TableData, columnIndex: number, ignoreNulls: boolean, nullAsZero: boolean) => ColumnStats;
+
+export interface StatCalculatorInfo {
+  value: string; // The ID - value maps directly to select component
+  label: string; // The name - label for Select component
+  description: string;
+  alias?: string; // optional secondary key.  'avg' vs 'mean', 'total' vs 'sum'
+
+  // Internal details
+  emptyInputResult?: any; // typically null, but some things like 'count' & 'sum' should be zero
+  standard: boolean; // The most common stats can all be calculated in a single pass
+  calculator?: StatCalculator;
+}
+
+/**
+ * @param ids list of stat names or null to get all of them
+ * @param notFound optional error object that will be filled with the names on unknown stats
+ */
+export function getStatsCalculators(ids?: string[], notFound?: string[]): StatCalculatorInfo[] {
+  if (ids === null || ids === undefined) {
+    return listOfStats;
+  }
+  return ids.reduce((list, id) => {
+    const stat = getById(id);
+    if (stat) {
+      list.push(stat);
+    } else if (notFound && id) {
+      notFound.push(id);
+    }
+    return list;
+  }, new Array<StatCalculatorInfo>());
+}
+
+export interface CalculateStatsOptions {
+  data: TableData;
+  columnIndex: number;
+  stats: string[]; // The stats to calculate
```

Comment:
Probably minor thing. Since this allows arbitrary strings, it can result in `notFound` list for unknown stats. I'm not sure why `notFound` list would be useful, as we are calculating stats on series from the same domain (this is my assumption, is it correct?). 

Would it be possible to get rid of `notFound` list and simply allow known stats only here? `stats: StatId[]`

Reply:
Oh, maybe a see a point of that now. So, for plugins, take Piechart as an example, there might be a need for custom stats, take `percentage` as an example (https://github.com/grafana/piechart-panel/blob/master/src/legend.js#L187). Hence, we could introduce extension point for plugins to enable custom stats calculation, i.e.:
```ts
reactPanel.setCustomStatsCalculator(statName: string, statCalculator: (allSeries, series) => number)
```

`setCustomStatsCalculator` would hook into our defaul stats calculator and extend it with custom calc that would be interesting from particular plugin POV.

Would that make sense @ryantxu @torkelo?

Reply:
My thinking with `notFound` was just for UI validation/feedback.  But looking again, we could easily just check that a key is in the output so I think we could drop that.

> reactPanel.setCustomStatsCalculator

I don't think it needs to be on the panel -- the panel already has control and could do whatever special calcs it needs.  

My thinking on having the calculator index is in response to the current design (in master) that calculates all stats all the time regardless of if they are needed/used.  The goal here is to:
1. let it work on table columns
2. only calculate when asked for
3. let it do more efficient calcs if possible (ie first=rows[0][columnIndex])

I am not worried about adding new functions now, but we should write it in a way that that is not hard later :)

Reply:
I removed the `notFound` parameter and check errors manually in the StatsPicker component